examples/plot_metadata_routing.py

```diff
@@ -0,0 +1,637 @@
+#                  # which is expected by the sub-estimator
+#             self.estimator_.fit(X, y, sample_weight=aliased_sample_weight)
+#    ...
+
+# %%
+# Router and Consumer
+# -------------------
+# To show how a slightly more complicated case would work, consider a case
+# where a meta-estimator uses some metadata, but it also routes them to an
+# underlying estimator. In this case, this meta-estimator is a consumer and a
+# router at the same time. This is how we can implement one, and it is very
+# similar to what we had before, with a few tweaks.
+
+
+class RouterConsumerClassifier(MetaEstimatorMixin, ClassifierMixin, BaseEstimator):
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def fit(self, X, y, sample_weight, **fit_params):
+        if self.estimator is None:
+            raise ValueError("estimator cannot be None!")
+
+        check_metadata(self, sample_weight=sample_weight)
+
+        if sample_weight is not None:
+            fit_params["sample_weight"] = sample_weight
+
+        # meta-estimators are responsible for validating the given metadata
+        request_router = get_routing_for_object(self)
+        request_router.validate_metadata(params=fit_params, method="fit")
+        # we can use provided utility methods to map the given metadata to what
+        # is required by the underlying estimator
+        params = request_router.route_params(params=fit_params, caller="fit")
+        self.estimator_ = clone(self.estimator).fit(X, y, **params.estimator.fit)
+        self.classes_ = self.estimator_.classes_
+        return self
+
+    def predict(self, X, **predict_params):
+        check_is_fitted(self)
+        # same as in ``fit``, we validate the given metadata
+        request_router = get_routing_for_object(self)
+        request_router.validate_metadata(params=predict_params, method="predict")
+        # and then prepare the input to the underlying ``predict`` method.
+        params = request_router.route_params(params=predict_params, caller="predict")
+        return self.estimator_.predict(X, **params.estimator.predict)
+
+    def get_metadata_routing(self):
+        router = (
+            MetadataRouter(owner=self.__class__.__name__)
+            .add_self(self)
+            .add(estimator=self.estimator, method_mapping="one-to-one")
+        )
+        return router
+
+
+# %%
+# The key parts where the above estimator differs from our previous
+# meta-estimator is accepting ``sample_weight`` explicitly in ``fit`` and
+# including it in ``fit_params``. Making ``sample_weight`` an explicit argument
+# makes sure ``set_fit_request(sample_weight=...)`` is present for this class.
+#
+# In ``get_metadata_routing``, we add ``self`` to the routing using
+# ``add_self``. Now let's look at some examples:
+
+# %%
+# - No metadata requested
+est = RouterConsumerClassifier(estimator=ExampleClassifier())
+print_routing(est)
+
+
+# %%
```

Comment:
I thinkt that we should give more information regarding what happen with the `add_self`. Once we print the metatadata tree, we can mention that we have a new key `$self` and it corresponds to the consumer behaviour of the meta-estimator.

Reply:
Added some info.